juice/src/net/container/sequential.rs

```diff
@@ -0,0 +1,317 @@
+use std::collections::HashMap;
+use std::fmt::{Debug, Formatter};
+
+use crate::co::IBackend;
+use crate::net::{layer_from_config, Context, Descriptor, Inout, Layer, LayerConfig};
+use crate::util::LayerOps;
+
+// Config for a single child layer.
+#[derive(Debug, Clone, Default)]
+pub struct SequentialChildConfig {
+    name: String,
+    config: LayerConfig,
+    inputs: Vec<String>,
+    outputs: Vec<String>,
+}
+
+// Config for the sequential layer, listing child layers and (optionally) connections between them.
+#[derive(Debug, Clone, Default)]
+pub struct SequentialConfig {
+    // List of named inputs. If not given, only a single input is assumed.
+    inputs: Vec<String>,
+
+    // Sequence of layers.
+    layers: Vec<SequentialChildConfig>,
+
+    // List of named outputs. If not given, Sequential layer outputs will match the outputs
+    // of the last layer.
+    outputs: Vec<String>,
+}
+
+// A container layer that composes nested layers in a sequence.
+//
+// In the most simple case, all nested layers are executed one by one, with outputs of one layer
+// becoming inputs of the next one. Inputs of the Sequential layers are inputs to the first
+// nested layer, and outputs of the last nested layer are the outputs of the Sequential layer:
+//
+// ```
+// let mut cfg = SequentialConfig::new();
+// cfg.add_layer("linear", LayerConfig::Linear(LinearConfig{ output_size: 10}));
+// cfg.add_layer("softmax", LayerConfig::Softmax);
+// ```
+//
+// Sequential layer also supports complex flow graphs (as long as they are acyclic)
+// by allowing nested layers inputs and outputs to be mapped to named "buffers":
+//
+// ```
+// let mut cfg = SequentialConfig::new();
+// cfg.map_input("in");
+// cfg
+//   .add_layer("linear1", LayerConfig::Linear(LinearConfig{ output_size: 10}))
+//   .map_input("in")
+//   .map_output("linear1_out");
+// cfg
+//   .add_layer("linear2", LayerConfig::Linear(LinearConfig{ output_size: 10}))
+//   .map_input("in")
+//   .map_output("linear2_out");
+// cfg
+//   .add_layer(/*some 2-input 1-output layer*/)
+//   .map_input("linear1_out")
+//   .map_input("linear2_out")
+//   .map_output("out");
+// cfg.add_output("out");
+// ```
+//
+// Note that currently it is requires that layer i can only use outputs from
+// previous layers 0..i-1.
+pub struct Sequential<B: IBackend> {
+    // Outward-facing inputs, outputs and params.
+    descriptor: Descriptor,
+    // Nested layers.
+    children: Vec<Box<dyn Layer<B>>>,
+}
+
+impl SequentialChildConfig {
+    pub fn map_input(&mut self, name: &str) -> &mut Self {
+        self.inputs.push(name.to_string());
+        self
+    }
+    pub fn map_output(&mut self, name: &str) -> &mut Self {
+        self.outputs.push(name.to_string());
+        self
+    }
+}
+
+impl SequentialConfig {
+    pub fn new() -> Self {
+        SequentialConfig::default()
+    }
+
+    pub fn add_layer(&mut self, name: &str, child_config: LayerConfig) -> &mut SequentialChildConfig {
```

Comment:
`child_config: impl Into<LayerConfig>` would make for a bit of more ergonomic usage

Reply:
Nice! Done.